File: Backend/services/user_service.go
package services

import (
	"AI-Powered-Automated-Loan-Underwriting-System/repositories"
	"context"
	"errors"
	"fmt"
	_ "time"

	"google.golang.org/protobuf/types/known/timestamppb"

	"golang.org/x/crypto/bcrypt"

	pb "AI-Powered-Automated-Loan-Underwriting-System/created_proto/user"
	"AI-Powered-Automated-Loan-Underwriting-System/models"
	"AI-Powered-Automated-Loan-Underwriting-System/utils"
)

type UserService struct {
	pb.UnimplementedUserServiceServer
	//DB   *gorm.DB
	repo *repositories.UserRepo
}

func NewUserService(repo *repositories.UserRepo) *UserService {
	return &UserService{repo: repo}
}

// Register user
func (s *UserService) Register(ctx context.Context, req *pb.RegisterRequest) (*pb.RegisterResponse, error) {
	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
	if err != nil {
		return nil, errors.New("failed to hash password")
	}

	user := models.User{
		FullName:    req.FullName,
		Email:       req.Email,
		Password:    string(hashedPassword),
		Phone:       req.Phone,
		DateOfBirth: req.DateOfBirth.AsTime(), // Convert string to time if needed
		Address:     req.Address,
	}

	if err := s.repo.CreateUser(ctx, user); err != nil {
		return nil, errors.New("failed to create user")
	}

	return &pb.RegisterResponse{
		Message: "User registered successfully",
		Status:  200,
	}, nil
}

// Login user
func (s *UserService) Login(ctx context.Context, req *pb.LoginRequest) (*pb.LoginResponse, error) {
	var user models.User
	if err := s.repo.GetUserByEmail(ctx, req, &user); err != nil {
		return nil, errors.New("invalid email or password")
	}

	if err := bcrypt.CompareHashAndPassword([]byte(user.Password), []byte(req.Password)); err != nil {
		fmt.Println("Password mismatch:", err)
		return nil, errors.New("invalid email or password")
	}

	token, err := utils.GenerateJWT(user.ID, user.Email)
	if err != nil {
		return nil, errors.New("failed to generate token")
	}

	return &pb.LoginResponse{
		Token:  token,
		Status: 200,
	}, nil
}

// GetCurrentUser retrieves the current user based on the token
func (s *UserService) GetUserDetails(ctx context.Context, req *pb.UserDetailsRequest) (*pb.UserDetailsResponse, error) {
	// var userID any = ctx.Value(middleware.ContextUserIDKey)
	// fmt.Println(userID)

	var userID any = req.UserId

	var user models.User
	if err := s.repo.GetUserByID(ctx, userID.(uint), &user); err != nil {
		return nil, errors.New("user not found")
	}

	var dob *timestamppb.Timestamp
	if !user.DateOfBirth.IsZero() {
		dob = timestamppb.New(user.DateOfBirth)
	} else {
		dob = nil
	}
	fmt.Println(user.DateOfBirth)

	return &pb.UserDetailsResponse{
		FullName:    user.FullName,
		Email:       user.Email,
		Phone:       user.Phone,
		Address:     user.Address,
		DateOfBirth: dob,
		CreditScore: int32(user.CreditScore),
		Status:      200,
	}, nil
}

// GetCreditScore retrieves the credit score of a user
func (s *UserService) GetUserCreditScore(ctx context.Context, req *pb.UserCreditScoreRequest) (*pb.UserCreditScoreResponse, error) {
	var user models.User
	if err := s.repo.GetUserByID(ctx, uint(req.UserId), &user); err != nil {
		return nil, errors.New("user not found")
	}

	if err := s.repo.GetCreditScore(ctx, uint(req.UserId), &user); err != nil {
		return nil, errors.New("failed to retrieve credit score")
	}
	fmt.Println(user)

	return &pb.UserCreditScoreResponse{
		CreditScore: int32(user.CreditScore),
		Status:      200,
	}, nil
}

// UpdateUserDetails updates the user info
func (s *UserService) UpdateUserDetails(ctx context.Context, req *pb.UserUpdateRequest) (*pb.UserUpdateResponse, error) {
	var user models.User
	if err := s.repo.GetUserByID(ctx, uint(req.UserId), &user); err != nil {
		return nil, errors.New("user not found")
	}

	user.FullName = req.FullName
	user.Phone = req.Phone
	user.DateOfBirth = req.DateOfBirth.AsTime()
	user.Address = req.Address

	if err := s.repo.UpdateUser(ctx, &user); err != nil {
		return nil, errors.New("failed to update user details")
	}

	return &pb.UserUpdateResponse{
		Message: "User details updated successfully",
		Status:  200,
	}, nil
}

File: Frontend/src/services/UserService.ts
import { userClient } from './Grpc.ts';
import {
    RegisterRequest,
    //RegisterResponse,
    LoginRequest,
    // LoginResponse,
} from '../proto/user';
import { Timestamp } from "../proto/google/protobuf/timestamp"; // adjust path if needed

export interface RegisterUserData {
    full_name: string;
    email: string;
    password: string;
    phone: string;
    date_of_birth: string;
    address: string;
}

export interface LoginUserData {
    email: string;
    password: string;
}

export interface RegisterResponseData {
    message: string;
    status: number;
}

export interface LoginResponseData {
    token: string;
    status: number;
}

function toTimestamp(dateStr: string): Timestamp {
  const ms = Date.parse(dateStr);
  if (isNaN(ms)) {
    throw new Error(`Invalid date: ${dateStr}`);
  }

  return {
    seconds: Math.floor(ms / 1000), // ✅ keep it as number
    nanos: (ms % 1000) * 1e6,
  };
}


export const registerUser = async (
    data: RegisterUserData
  ): Promise<RegisterResponseData> => {
    const request: RegisterRequest = {
      fullName: data.full_name,
      email: data.email,
      password: data.password,
      phone: data.phone,
      dateOfBirth: toTimestamp(data.date_of_birth), // ✅ now this works
      address: data.address,
    };
  
    const response = await userClient.Register(request);
  
    return {
      message: response.message,
      status: response.status,
    };
  };
  


export const loginUser = async (
    data: LoginUserData
): Promise<LoginResponseData> => {
    const request: LoginRequest = {
        email: data.email,
        password: data.password,
    };

    const response = await userClient.Login(request);

    return {
        token: response.token,
        status: response.status,
    };
};